mesonbuild/build.py

```diff
@@ -1454,11 +1454,20 @@ def link_whole(self, target):
                     raise InvalidArguments(msg + ' This is not possible in a cross build.')
                 else:
                     mlog.warning(msg + ' This will fail in cross build.')
-            if isinstance(self, StaticLibrary):
-                # When we're a static library and we link_whole: to another static
-                # library, we need to add that target's objects to ourselves.
+            # When we're a static library and we link_whole: to another static
+            # library, we need to add that target's objects to ourselves.
+            # Otherwise add it to the link_whole_targets, but not both.
+            #
+            # We cannot use this with rust because of the way rust does crate
+            # search, or with fortran because of modules.
+            # TODO: it's like C++ with modules will not be able to suppor this
+            # optimization either
```

Comment:
link_whole != link_with

I don't think this is an optimization? It's just necessary.

A fundamental aspect of whole inclusion of one static library in another is that the entire contents must be inside the second static library. We cannot pass the static archive as an option to ar, because that would mean that one `*.a` file contains a bunch of `*.o` files plus another `*.a` file, and it doesn't really make sense -- or work -- to have an archive inside of another archive.

What does this mean for rust/fortran/C++modules and link_whole?

Reply:
For both Rust and Fortran, you must have a module named "foo" if you've called `extern foo` or `include foo` (respectively). If you take foo and extract it's objects and shove it inside of `bar` because you want to consume both `foo and bar` then you'll fail to link because you won't have a `libfoo.a`.

Reply:
Right, so I'd assume in this case that you ***must*** not use link_whole (it's an error to do it) and must instead use link_with.

Reply:
I thought we had an optimization where we *didn't* link at all for convenience libraries, we just say link final targets with both `liba` and `libb`.

Reply:
I'm not actually sure what that means.

Convenience libraries don't need to link anything, it's not an optimization, it's just fundamentally how things work.

```meson
static_library('bar', 'bar.c', install: true)
static_library('foo', 'foo.c', link_with: libbar_static, install: true)
```
libfoo must not include libbar's objects in an installable product, because this is a logical error, linking two copies of one object (foo.pc will include both `-lfoo` and `-lbar`).

If libbar is `install: false` then originally, exactly the same thing happened, except that -lbar didn't work because there wasn't one, and the universe imploded.

if libfoo is also `install: false` then no one cares, there's no pkg-config and nothing to install.

..

In-tree dependency structures behaved exactly the same as pkg-config, because that's the point of dependency tracking. `link_with: libfoo` would add both libfoo and libbar to the command line, and all is well.

...

The logic was overall buggy for case 2, so a bug fix was made that causes the combination of `link_with: uninstalled_thing, install: true` to magically rewrite your build and pretend you wrote what you obviously meant, which is `link_whole`.

Reply:
There was a separate bug, IIRC, in which `link_whole: static_library(...)` simply passed libbar.a as a parameter to `ar csrD`, so libfoo.a's archive contents were "foo.c.o" and "libbar.a", which was another kind of broken.

Reply:
This PR has literally nothing to do with link_with, and everything to do with link_whole. And again, there's no optimization going on, now or in the past.

And the point of `static_library(..., link_whole: ...)` is that you're explicitly saying you want objects, so if you try doing that with rust or fortran then what?

You're being silly for using the `link_whole` kwarg, it sounds like, and you should absolutely not be doing that?

What is the semantic meaning of "libfoo", a rust static library, using `link_whole` on another rust static library "libbar"? Why did someone decide to do this? Were they crazy? Foolish? Just didn't understand rust on a conceptual level? We should tell them that that's an error, and bail out, rather than rewriting the semantics to `link_with`.

Reply:
Fine, I'm not going to fight about whether this is an optimization, I'll change the commit message + comment.

Rust doesn't link_with two static libraries, ever. It *always* link_whole's static libraries (both C ABI "staticlib" and Rust ABI "rlib"), and then rustc relies on lto to drop unnecessary symbols at executable/shared-library link time, and for each crate that's imported there must be an `--extern` switch given. You can look at `test cases/rust/17 staticlib link static` to see what I mean.

Reply:
It's not the wording I'm worried about. :) Just trying to understand whether, based on the stated purposes of these kwargs, it makes sense to allow link_whole or take into account people using it "the wrong way".

From what you're saying it sounds like people should never, ever use link_whole with rust, because rust doesn't really have that concept at all. You just link "with", and all recursive rlib/staticlibs get included.

Perhaps instead of this condition we should add an interpreter error: "cannot pass rust libraries in link_whole, use link_with instead"?

Reply:
Which is kinda problematic because at the moment we can't konw if a target is an rlib library, because the build module is such a mess. Sigh

Reply:
Why doesn't uses_rust() work?

Reply:
Oh wait, I see your edit now.

So the current logic in this PR for uses_rust() isn't fully accurate because it's possible to link_whole a rust staticlib (but not an rlib) and this PR would cause that to not work by instead propagating it as a dependency?

(This would produce incorrect results, I think, if you want to link_whole an uninstalled rust staticlib into an installed static library.)

Reply:
Okay. I've done some more reading and here's what should be possible:
- link_whole (which is not implemented in meson) shoud support C ABI libraries, either those compiled from C or those compiled with the rust crate type of `staticlib`
- link_with should accept Rust ABI `rlib`'s, it must not include the objects and instead use `--externa foo=...`
- link_with should accept C ABI `staticlib` or non-rust targets and include them with `-l static=foo`

I'll send a seperate PR to deal with link_whole, but, for the moment I think this can all be realatively easily be fixed enough for the general case.

Reply:
Which means the problem (for Rust) at least, is more deep rooted. We don't handle extracted objects from C libraries at all, and we fail to set up proper dependencies... sigh